Fix convolution edge behavior with fill=nan

keflavich opened the pull request:
This is work to fix issue #16497, in which it was reported that there is incorrect behavior in the treatment of edges with non-fft-based convolution.

This is a WIP - the progress so far is just implementing a regression test.  The solution is not yet clear; a workaround is to do a weighted convolution like the FFT version does, but I think that's the wrong solution as it doubles the computation time.  Part of the solution may be to remap `boundary='fill', fill=nan` to a different boundary type, but attempting that revealed that `boundary=None` is also bugged - it bleeds in zeros from the edge that should simply not happen.

Fixes #16497

keflavich commented:
EDIT : related to #11920 but does not fix (see Larry's comment below)

pllim commented:
Need change log. Thanks!

keflavich commented:
There is a changelog entry - why is it complaining that there is not?

larrybradley commented on docs/changes/convolution/18348.api.rst:
```diff
@@ -0,0 +1,2 @@
+Fixes edge treatment in ``convolve(..., boundary='fill', fill=nan)`` to be consistent with
```
```suggestion
Fixes edge treatment in ``convolve(..., boundary='fill', fill_value=nan)`` to be consistent with
```

larrybradley reviewed (changes requested):
@keflavich This fix looks fine (except for a small nit in the changelog snippet).  It fixes, #16497, but this PR has no effect on #11920.  The results for the example in #11920 are the same with and without this PR, so perhaps remove the "Fixes #11920" comment.

IMHO, the convolve results for #11920 where an all-NaN input array returns all NaN is more what I expected -- there is not valid data to interpolate.  The result where `convolve_fft` returns an array of all zeros appears to be a side effect of replacing all NaN values to 0 before the FFT.

keflavich committed:
add a regression test for issue 16497;  fix a testing issue - cannot convolve_fft with boundary=None. Fix the original issue.

merge: tests added by both branches

merge: more testing

merge: test change

add changelog entry

keflavich committed:
fix fill->fill_value

keflavich commented:
I tried removing the 'fixes' comment - is that going to auto-update or do I need to do something else to unlink 11920?

larrybradley reviewed (approved):
Thanks, @keflavich.  But we should keep #11920 open, since this PR has no effect for that issue.

larrybradley commented:
If you edit your initial comment and remove "fixes https://github.com/astropy/astropy/issues/11920 too", I think it should remove closing that issue.

larrybradley commented:
I just removed the comment and it worked.
